# Clasificador euclideano determinista a priori

In [2]:
import numpy as np # Cómputo numérico
from sympy import Matrix, symbols, log, simplify # Cómputo simbólico

In [3]:
def clasif_e(samples):
    # Vector genérico usando Sympy
    X = Matrix([symbols('x' + str(i+1)) for i in range(samples[0].shape[0])])
    print(' Variables : ', X)
    fds = []
    for s in samples:
        m = Matrix(np.mean(s, axis=1))
        fds.append(simplify(X.T*m - (m.T*m)/2))
    return fds

In [4]:
w1 = np.array([[1,5,6,1],[2,3,4,-2],[0,1,5,3],[1,3,5,2]]).T
w2 = np.array([[6,8,-1,6],[7,9,3,6],[8,7,1,9]]).T
samples = (w1,w2)
fds = clasif_e(samples)
print(fds)
display(fds[0])
display(fds[1])
for fd in range(len(fds)):
    print('fd' + str(fd+1) + ': ', fds[fd].evalf(subs={'x1':3, 'x2':1, 'x3':3, 'x4':1}))

 Variables :  Matrix([[x1], [x2], [x3], [x4]])
[Matrix([[1.0*x1 + 3.0*x2 + 5.0*x3 + 1.0*x4 - 18.0]]), Matrix([[7.0*x1 + 8.0*x2 + 1.0*x3 + 7.0*x4 - 81.5]])]


Matrix([[1.0*x1 + 3.0*x2 + 5.0*x3 + 1.0*x4 - 18.0]])

Matrix([[7.0*x1 + 8.0*x2 + 1.0*x3 + 7.0*x4 - 81.5]])

fd1:  Matrix([[4.00000000000000]])
fd2:  Matrix([[-42.5000000000000]])


### Tarea

Obtener las funciones discriminantes para las muestras:

$$ \omega_1 =  \{(0.5, 10.5),(1, 12.5),(3, 10.5),(3, 12.5),(3, 14.5),(3, 18),(5, 18),(5, 16),(5, 14.5),(5, 13)\} $$

$$ \omega_2 = \{(6, 9),(8, 10),(9, 11),(8.5, 12),(7, 13.5),(8, 16)\} $$

Para calcular el centroide de las muestras, se calcula la media de cada fila.

$$ \omega_1 = \left\{ \begin{pmatrix} 0.5 \\ 10.5 \end{pmatrix},  \begin{pmatrix} 1 \\ 12.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 10.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 12.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 14.5 \end{pmatrix},\begin{pmatrix} 3 \\ 18 \end{pmatrix},\begin{pmatrix} 5 \\ 18 \end{pmatrix},\begin{pmatrix} 5 \\ 16 \end{pmatrix},\begin{pmatrix} 5 \\ 14.5 \end{pmatrix},\begin{pmatrix} 5 \\ 13 \end{pmatrix} \right\} $$

$$ \omega_2 = \left\{ \begin{pmatrix} 6 \\ 9 \end{pmatrix},  \begin{pmatrix} 8 \\ 10 \end{pmatrix}, \begin{pmatrix} 9 \\ 11 \end{pmatrix}, \begin{pmatrix} 8.5 \\ 12 \end{pmatrix}, \begin{pmatrix} 7 \\ 13.5 \end{pmatrix}, \begin{pmatrix} 8 \\ 16 \end{pmatrix} \right\} $$

Los centroides de las dos clases son:

In [5]:
omega1 = np.array([[0.5, 10.5],[1, 12.5],[3, 10.5],[3, 12.5],[3, 14.5],[3, 18],[5, 18],[5, 16],[5, 14.5],[5, 13]]).T
omega2 = np.array([[6, 9],[8, 10],[9, 11],[8.5, 12],[7, 13.5],[8, 16]]).T
z_1 = np.mean(omega1, axis=1)
z_2 = np.mean(omega2, axis=1)
print(z_1)
print(z_2)

[ 3.35 14.  ]
[ 7.75       11.91666667]


$$ z_1 = \left\{ \begin{pmatrix} 3.35 \\ 14 \end{pmatrix} \right\} $$

$$ z_2 = \left\{ \begin{pmatrix} 7.75 \\ 11.9166 \end{pmatrix} \right\} $$

Las funciones discrimantes son:

$$ fd_K(X) = X^T \cdot Z_K - \frac{1}{2} \cdot Z^T_K \cdot Z_K \quad \quad \forall k \in K  $$

$$ fd_1(X) = X^T \cdot \begin{pmatrix} 3.35 \\ 14 \end{pmatrix} - \frac{1}{2} \cdot \begin{pmatrix} 3.35 \\ 14 \end{pmatrix}^T \cdot \begin{pmatrix} 3.35 \\ 14 \end{pmatrix} $$

$$ fd_2(X) = X^T \cdot \begin{pmatrix} 7.75 \\ 11.9166 \end{pmatrix} - \frac{1}{2} \cdot \begin{pmatrix} 7.75 \\ 11.9166 \end{pmatrix}^T \cdot \begin{pmatrix} 7.75 \\ 11.9166 \end{pmatrix} $$

Desarrollando las operaciones:

$$ fd_1 (X) = 3.35 x_1 +14 x_2 - 103.611 $$

$$ fd_2(X) = 7.75 x_1 + 11.9166 x_2 - 101.033 $$

In [6]:
samples = (omega1, omega2)
fds = clasif_e(samples)
print(fds)

 Variables :  Matrix([[x1], [x2]])
[Matrix([[3.35*x1 + 14.0*x2 - 103.61125]]), Matrix([[7.75*x1 + 11.9166666666667*x2 - 101.034722222222]])]


In [7]:
# fd1
display(fds[0])
# fd2
display(fds[1])

Matrix([[3.35*x1 + 14.0*x2 - 103.61125]])

Matrix([[7.75*x1 + 11.9166666666667*x2 - 101.034722222222]])

Graficamente

# Clasificador con distancia de Mahalanobis

In [8]:
def clasif_m(samples):
    X = Matrix([symbols('x' + str(i+1)) for i in range(samples[0].shape[0])])
    print('Variables: ', X)
    fds = []
    for s in samples:
        m = Matrix(np.mean(s, axis=1))
        m_cov = np.cov(s, bias=True)
        # print(m_cov)
        m_ci = Matrix(m_cov).inv()
        fds.append(simplify((X.T*m_ci*X)/-2 + X.T*m_ci*m - (m.T*m_ci*m)/2 - Matrix([log(m_ci.det())/2])))
    return fds

In [9]:
w1 = np.array([[1,2],[2,2],[3,1],[2,3],[3,2]]).T
w2 = np.array([[8,10],[9,8],[9,9],[8,9],[7,9]]).T
samples = (w1,w2)
fds = clasif_m(samples)
print(fds)

Variables:  Matrix([[x1], [x2]])
[Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 6.95652173913044*x1 - 1.52173913043478*x2**2 + 8.47826086956522*x2 - 16.9768445432953]]), Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 27.6086956521739*x1 - 1.52173913043478*x2**2 + 36.304347826087*x2 - 277.411627151991]])]


In [10]:
display(fds[0])
display(fds[1])

Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 6.95652173913044*x1 - 1.52173913043478*x2**2 + 8.47826086956522*x2 - 16.9768445432953]])

Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 27.6086956521739*x1 - 1.52173913043478*x2**2 + 36.304347826087*x2 - 277.411627151991]])

### Tarea
Obtener las funciones discriminantes para las muestras:
$$ \omega_1 = \{ (0.5, 10.5),(1, 12.5),(3, 10.5),(3, 12.5),(3, 14.5),(3, 18),(5, 18),(5, 16),(5, 14.5),(5, 13) \} $$

$$ \omega_2 = \{ (6, 9),(8, 10),(9, 11),(8.5, 12),(7, 13.5) (8, 16) \} $$

**Nota:** Las matrices de covarianzas no son iguales: se obtendrán funciones discriminantes
cónicas.

$$ \omega_1 = \left\{ \begin{pmatrix} 0.5 \\ 10.5 \end{pmatrix},  \begin{pmatrix} 1 \\ 12.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 10.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 12.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 14.5 \end{pmatrix}, \begin{pmatrix} 3 \\ 18 \end{pmatrix}, \begin{pmatrix} 5 \\ 18 \end{pmatrix}, \begin{pmatrix} 5 \\ 16 \end{pmatrix}, \begin{pmatrix} 5 \\ 14.5 \end{pmatrix}, \begin{pmatrix} 5 \\ 13 \end{pmatrix} \right\} $$

$$ \omega_2 = \left\{ \begin{pmatrix} 6 \\ 9 \end{pmatrix},  \begin{pmatrix} 8 \\ 10 \end{pmatrix}, \begin{pmatrix} 9 \\ 11 \end{pmatrix}, \begin{pmatrix} 8.5 \\ 12 \end{pmatrix}, \begin{pmatrix} 7 \\ 13.5 \end{pmatrix}, \begin{pmatrix} 8 \\ 16 \end{pmatrix} \right\} $$

Las medias (el cálculo es análogo al del centroide) son:

In [12]:
m_1 = np.mean(omega1, axis=1)
m_2 = np.mean(omega2, axis=1)
print(m_1)
print(m_2)

[ 3.35 14.  ]
[ 7.75       11.91666667]


$$ \vec{m}_1 = \begin{pmatrix} 3.35 \\ 14 \end{pmatrix}$$

$$ \vec{m}_2 = \begin{pmatrix} 7.75 \\ 11.9166 \end{pmatrix}$$

Las matrices de covarianzas:

$$ c11 = \frac{1}{10} ((0.5-3.35)^2 + (1-3.35)^2 + (3-3.35)^2 + \cdots + (5-3.35)^2)$$ 
$$ c11 = 2.5025$$

$$ c12 = c21 = \frac{1}{10}((0.5-3.35)(10.5-14)+(1-3.35)(12.5-14)+(3-3.35)(10.5-14) + \cdots + (5-3.35)(13-14))$$
$$ c12 = 2.275 $$

$$ c22 = \frac{1}{10} ((10.5-14)^2 + (12.5-14)^2 + (10.5-14)^2 + \cdots + (13-14)^2)$$ 
$$ c22 = 6.65 $$

$$ C_1 = \begin{pmatrix}
2.5025 & 2.275 \\
2.275 & 6.65 \\
\end{pmatrix} $$

In [14]:
cov_1 = np.cov(omega1, bias=True)
cov_1

array([[2.5025, 2.275 ],
       [2.275 , 6.65  ]])

---

$$ c11 = \frac{1}{10} ((6-7.75)^2 + (8-7.75)^2 + (9-7.75)^2 + \cdots + (8-7.75)^2)$$ 
$$ c11 = 0.979166$$

$$ c12 = c21 = \frac{1}{10}((6-7.75)(9-11.9166) + (8-7.75)(10-11.9166) + (9-7.75)(11-11.9166) + \cdots + (8-7.75)(16-11.9166))$$
$$ c12 = 0.562499 $$

$$ c22 = \frac{1}{10} ((9-11.9166)^2 + (10-11.9166)^2 + (11-11.9166)^2 + \cdots + (16-11.9166)^2)$$ 
$$ c22 = 5.368055 $$

$$ C_2 = \begin{pmatrix}
0.979166 & 0.562499 \\
0.562499 & 5.368055 \\
\end{pmatrix} $$

In [16]:
cov_2 = np.cov(omega2, bias=True)
cov_2

array([[0.97916667, 0.5625    ],
       [0.5625    , 5.36805556]])

Las inversas de las matrices de covarianzas:

In [17]:
cov_1_inv = np.linalg.inv(cov_1)
cov_2_inv = np.linalg.inv(cov_2)
print(cov_1_inv)
print(cov_2_inv)

[[ 0.57997558 -0.1984127 ]
 [-0.1984127   0.21825397]]
[[ 1.08669166 -0.11387067]
 [-0.11387067  0.19821931]]


$$ C^{-1}_1 =  \begin{pmatrix} 0.57997558 & -0.1984127 \\ -0.1984127 & 0.21825397 \\ \end{pmatrix} $$

$$ C^{-1}_2 =  \begin{pmatrix} 1.08669221 & -0.11387053 \\ -0.11387053 & 0.19821929 \\ \end{pmatrix} $$

Los determinantes:

In [19]:
det_c1 = np.linalg.det(cov_1)
det_c2 = np.linalg.det(cov_2)
print(det_c1)
print(det_c2)

11.466
4.939814814814815


$$ |C_1| = 11.466$$

$$ |C_2| = 4.939814814814815$$


Las funciones discriminantes son:

$$ fd_i(X) = -\frac{1}{2} X^T \cdot C^{-1}_i \cdot X + X^T \cdot C^{-1}_i \cdot \vec{m}_i - \frac{1}{2} \vec{m}^T_i \cdot C^{-1}_i \cdot \vec{m}_i - \frac{1}{2} \ln |C_i| $$

$$ fd_1(X) = -\frac{1}{2} X^T \cdot C^{-1}_1 \cdot X + X^T \cdot C^{-1}_1 \cdot \vec{m}_1 - \frac{1}{2} \vec{m}^T_1 \cdot C^{-1}_1 \cdot \vec{m}_1 - \frac{1}{2} \ln |C_1| $$

$$ fd_1  = -1.08695652173913 x_1^2 - 1.08695652173913 x_1x_2 + 6.95652173913044 x_1 - 1.52173913043478 x_2^2 + 8.47826086956522 x_2 - 16.9768445432953$$

$$ fd_2(X) = -\frac{1}{2} X^T \cdot C^{-1}_2 \cdot X + X^T \cdot C^{-1}_2 \cdot \vec{m}_2 - \frac{1}{2} \vec{m}^T_2 \cdot C^{-1}_i \cdot \vec{m}_2 - \frac{1}{2} \ln |C_2| $$

$$ fd_2 = -1.08695652173913 x_1^2 - 1.08695652173913 x_1x_2 + 27.6086956521739 x_1 - 1.52173913043478 x_2^2 + 36.304347826087 x_2 - 277.411627151991$$

In [20]:
samples = (omega1,omega2)
fds = clasif_m(samples)
print(fds)

Variables:  Matrix([[x1], [x2]])
[Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 6.95652173913044*x1 - 1.52173913043478*x2**2 + 8.47826086956522*x2 - 16.9768445432953]]), Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 27.6086956521739*x1 - 1.52173913043478*x2**2 + 36.304347826087*x2 - 277.411627151991]])]


In [21]:
display(fds[0])
display(fds[1])

Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 6.95652173913044*x1 - 1.52173913043478*x2**2 + 8.47826086956522*x2 - 16.9768445432953]])

Matrix([[-1.08695652173913*x1**2 - 1.08695652173913*x1*x2 + 27.6086956521739*x1 - 1.52173913043478*x2**2 + 36.304347826087*x2 - 277.411627151991]])